src/librustc_mir/build/expr/as_rvalue.rs

```diff
@@ -59,13 +59,38 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
 
         match expr.kind {
             ExprKind::Scope { extent, value } => {
+                let extent = (extent, source_info);
                 this.in_scope(extent, block, |this| this.as_rvalue(block, scope, value))
             }
             ExprKind::Repeat { value, count } => {
                 let value_operand = unpack!(block = this.as_operand(block, scope, value));
                 block.and(Rvalue::Repeat(value_operand, count))
             }
             ExprKind::Borrow { region, borrow_kind, arg } => {
+                if let Region::ReScope(extent) = *region {
+                    // borrows of regions outside closures can
+                    // occur. For such cases, we simply let the region
+                    // extend past the end of the closure.
+                    //
+                    // (Also, in principle `extent` should occur at
+                    // most once on scopes, but that invariant does
+                    // not currently hold.)
+
+                    // find the scope associated with this extent, and
```

Comment:
I don't mind trying to do it that way. I can gather data on how many extraneous `EndRegion` statements are emitted and then removed.